src/services/refactors/inlineVariable.ts

```diff
@@ -0,0 +1,194 @@
+
+const inlineVariableAction = {
+    name: refactorName,
+    description: refactorDescription,
+    kind: "refactor.inline.variable"
+};
+
+interface InliningInfo {
+    references: Node[];
+    declaration: VariableDeclaration;
+    replacement: Expression;
+}
+
+registerRefactor(refactorName, {
+    kinds: [ inlineVariableAction.kind ],
+
+    getAvailableActions(context) {
+        const {
+            file,
+            program,
+            preferences,
+            startPosition,
+            triggerReason
+        } = context;
+
+        // tryWithReferenceToken is true below when triggerReason === "invoked", since we want to
+        // always provide the refactor in the declaration site but only show it in references when
+        // the refactor is explicitly invoked.
+        const info = getInliningInfo(file, startPosition, triggerReason === "invoked", program);
+        if (!info) {
+            return emptyArray;
+        }
+
+        if (refactor.isRefactorErrorInfo(info) && preferences.provideRefactorNotApplicableReason) {
+            return [{
+                name: refactorName,
+                description: refactorDescription,
+                actions: [{
+                    ...inlineVariableAction,
+                    notApplicableReason: info.error
+                }]
+            }];
+        }
+
+        if (!refactor.isRefactorErrorInfo(info)) {
+            return [{
+                name: refactorName,
+                description: refactorDescription,
+                actions: [inlineVariableAction]
+            }];
+        }
+
+        return emptyArray;
+    },
+
+    getEditsForAction(context, actionName) {
+        Debug.assert(actionName === refactorName, "Unexpected refactor invoked");
+
+        const { file, program, startPosition } = context;
+
+        const info = getInliningInfo(file, startPosition, /*tryWithReferenceToken*/ true, program);
+        if (!info || refactor.isRefactorErrorInfo(info)) {
+            return undefined;
+        }
+
+        const { references, declaration, replacement } = info;
+        const edits = textChanges.ChangeTracker.with(context, tracker => {
+            for (const node of references) {
+                tracker.replaceNode(file, node, getReplacementExpression(node, replacement));
+            }
+            tracker.delete(file, declaration);
+        });
+
+        return { edits };
+    }
+});
+
+function getInliningInfo(file: SourceFile, startPosition: number, tryWithReferenceToken: boolean, program: Program): InliningInfo | RefactorErrorInfo | undefined {
+    const checker = program.getTypeChecker();
+    const token = getTokenAtPosition(file, startPosition);
+    const parent = token.parent;
+
+    // If the node is a variable declaration, make sure it's not in a catch clause or for-loop
+    // and that it has a value.
+    if (isInitializedVariable(parent) && isVariableDeclarationInVariableStatement(parent) && isIdentifier(parent.name)) {

```

Comment:
Can you add a comment for which case this is supposed to capture? The below cases seem like they handle similar logic.

Reply:
What exactly would you like me to comment? I tried to explain what each `if` statement does, except for the ones in lines 137-152 since they're basically the same as the checks in the above if statement.

Reply:
Usually what I have in mind for these are snippets of example code.